### DISCUSSION TITLE
Привязка телеграмма к определенному пользователю

### DISCUSSION THREAD

**Discussion by Re-Requeste** [totumonline/totum-ru-issues-and-discussions#532] [Как это сделать?] [answered]
Всем привет. 
Необходимо было чтоб к каждому пользователю приходили  оповещение в Telegram через бота, ознакомился с [Создаем и подключаем Telegram бота](https://ru.docs.totum.online/telegram-bot) , все работает.
Но мне не понравилась что веб хук ловит абсолютно все сообщение, и загружает все в таблицу,  ведь я не делаю чат бота. 
Подумав и выстроив логику, то добился некоторых успехов, пользователь переходит по подготовленной ссылке в виде : t.me/ваш бот?start=сгенерированное значение , далее нажимает на кнопку старт, возвращается на страницу, нажимает на кнопку проверить подключение, тотум отправляет запрос getUpdates вашему боту, ищет значение которое вы передали пользователю и вытаскивает id пользователя.

Реализовал так, может кому пригодится ну или мне в будущем

![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/137899742/0128f9d3-064e-4bb6-8ad4-e3ab173b8001)

**Comment by LostAlly** [comment]
А почему бы код не вставить тегом код ну или хотя бы просто текстом, вы себе в будущем не скажете спасибо за такой подход )))

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Норм идея. Некоторые улучшения могу предложить тк делал эту штуку, только первый шаг вручную был. Те пользователь получал уникальный для него номер и его отправлял в Бот. Бот слушал на предмет именно номера. Если номер совпадал, то отправляли пользователю сообщение о том, что все ок. Ну и дальше отправляем ему сообщения.

Отправлять сообщения можно кстати функцией [notificationSend](https://ru.docs.totum.online/functions#fn-notificationSend). Там есть параметр custom, как раз для отправки ботам (он будет в паралельном процессе отправлять, чтоб основной не останавливать).

### Давайте ща пример более автоматический запилим (по этой теме с передачей /start сразу):

Делаем в таблице users поля строковые t_connect_code и t_room

> замените YOUR_BOT_NAME

```
= : linkToDataHtml(title: "Номер подключения к боту"; html: $if)

if: if(condition: $room != ""; then: "Этот аккаунт уже подключен к боту!"; else: $link)
	room: select(table: 'users'; field: 't_room'; where: 'id' = $#nu)


link: str`"Перейдите по ссылке и откройте Telegram на предложение браузера: <a href='https://t.me/YOUR_BOT_NAME?start=" + $if_num + "'>@YOUR_BOT_NAME"`

	if_num: if(condition: $select_num = ""; then: $while; else: $select_num)
	
		~select_num: select(table: 'users'; field: 't_connect_code'; where: 'id' = $#nu)
	
	while: while(action: $generate; action: $set; action: $generate)
	
		~generate: strRandom(length: 5; numbers: "true"; letters: "false"; symbols: "false")
		set: set(table: 'users'; field: 't_connect_code' = $generate; where: 'id' = $#nu; log: true)
```

Ниче страшного, что у вас ремоут слушает бота. Можно просто ничего лишнего в БД не писать. Пример в доках сделан с записью, что бы можно быстрее понять, что происходит.

Итак ремоут, на который смотрит бот:

```
=: if(condition: $#headers["X-Telegram-Bot-Api-Secret-Token"][0] = @telegram.h_secret; then: $action)

action: if(condition: $exist != ""; then: $send_exist; else: $start)
		send_exist: getFromScript(uri: str`"https://api.telegram.org/bot" + @telegram.h_token + "/" + $command`; post: "chat_id" = $chat_id; post: "text" = "Ваш аккаунт уже подключен!")
			exist: select(table: 'users'; field: 'id'; where: 't_room' = $chat_id)
	
	start: if(condition: $sel_user != ""; then: $add; else: $send_error)
			~sel_user: select(table: 'users'; field: 'id'; where: 't_connect_code' = $text_cort; where: 't_connect_code' != "")
		add: while(action: $set; action: $send_succses)
				set: set(table: 'users'; field: 't_room' = $chat_id; where: 'id' = $sel_user; log: true)
			send_succses: getFromScript(uri: str`"https://api.telegram.org/bot" + @telegram.h_token + "/" + $command`; post: "chat_id" = $chat_id; post: "text" = str`"Вы успешно подключены к пользователю" ++ $username`)
				username: select(table: 'users'; field: 'fio'; where: 'id' = $sel_user)
			send_error: getFromScript(uri: str`"https://api.telegram.org/bot" + @telegram.h_token + "/" + $command`; post: "chat_id" = $chat_id; post: "text" = str`"Некорректный код подключения" ++ $text_cort`)
	

~chat_id: $extract[message][chat][id]

~text_cort: strPart(str: $text; offset: 7)
	text: $extract[message][text]
		~extract: jsonExtract(text: $#input)

~command: "sendMessage"
```

+ text_cort — тк сообщение приходит /start 12345 и надо отступить 7 символов, что бы читать код подключения

+ @telegram.h_token — это у меня ключ лежит в таблице telegram в хедере в поле h_token

+ @telegram.h_secret — это у меня секрет лежит в таблице telegram в хедере в поле h_token

**Reply by Re-Requeste** [reply]
Вы так легко это описали, ужасно что  я не додумался про это раньше и долго ковырял код...
Вы гений, прекрасное решение....

**Reply by ttm-alexeypolunin** [reply]
Спасибо вам, что написали, тк мы не знали, что можно сразу бота вызвать по ссылке (у них же там доки у телеграма длиной в 10 колометров :) )
